others/serialization.cpp

```diff
@@ -0,0 +1,190 @@
+    }
+
+    /**
+     * Serializes a string to a binary file.
+     * @param out The output stream (std::ofstream).
+     * @param data The string to be serialized.
+     *
+     * The string is serialized by first storing its length, followed by the
+     * content.
+     */
+    static void serialize(std::ofstream &out, const std::string &data) {
+        ssize_t length = data.size();
+        serialize(out, length);           // Serialize the length of the string.
+        out.write(data.c_str(), length);  // Serialize the string characters.
+        out.put('|');  // Add a delimiter to denote the end of the string.
+    }
+};
+
+/**
+ * @class Deserializer
+ * A utility class for deserializing data from a binary file.
+ */
+class Deserializer {
+ public:
+    /**
+     * Deserializes fundamental data types (like int, float, double, etc.) from
+     * a binary file.
+     * @tparam T The type of the data to be deserialized.
+     * @param in The input stream (std::ifstream).
+     * @param data The variable where the deserialized data will be stored.
+     *
+     * @note This function only works for fundamental types (primitives).
+     */
+    template <typename T>
+    static void deserialize(std::ifstream &in, T &data) {
+        static_assert(
+            std::is_fundamental<T>::value,
+            "Non-fundamental types are not allowed for this function!");
+        in.read(reinterpret_cast<char *>(&data), sizeof(T));
+    }
+
+    /**
+     * Deserializes a string from a binary file.
+     * @param in The input stream (std::ifstream).
+     * @param data The string where the deserialized content will be stored.
+     *
+     * The string is deserialized by reading its length, followed by the
+     * characters, and finally validating the delimiter throws error if the
+     * delimiter '|' is not found.
+     */
+    static void deserialize(std::ifstream &in, std::string &data) {
+        ssize_t length;
+        deserialize(in, length);  // Deserialize the length of the string.
+
+        if (length > 1024 * 1024)  // Sanity check to prevent huge strings.
+        {
+            throw std::runtime_error(
+                "Deserialized string length is too large.");
+        }
+
+        data.resize(length);  // Resize the string to fit the incoming data.
+        in.read(&data[0], length);  // Read the string characters.
+        char delimiter;
+        in.get(delimiter);  // Check the delimiter.
+        if (delimiter != '|') {
+            throw std::runtime_error("Delimiter '|' not found after string.");
+        }
+    }
+};
+
+void runTests();
+
+/**
+ * Demonstrates the use of Serializer and Deserializer for fundamental types and
+ * strings.
+ */
+int main() {
+    std::ofstream outFile("output.bin", std::ios::binary);
+    std::ifstream inFile("output.bin", std::ios::binary);
+
+    if (!outFile || !inFile) {
+        std::cerr << "Error opening files.\n";
+        return 1;
```

Comment:
main function should have nothing except 
a call to the test function
and a return 0;
thus move all the tests from here to the test function